Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc

```diff
@@ -37,146 +36,182 @@ GEMGeometryBuilderFromCondDB::build(const std::shared_ptr<GEMGeometry>& theGeome
-    theGeometry->add(gep);
+  for( unsigned int id = 0; id < detids.size(); ++id ){  
+    GEMDetId gemid( detids[id] );
+    LogDebug("GEMGeometryBuilderFromDDD") <<"GEMGeometryBuilderFromDDD adding " << gemid << std::endl;
     
-    std::list<GEMEtaPartition *> gepList;
-    if( m_chids.find( chid ) != m_chids.end()) {
-      gepList = m_chids[chid];
-    }
-    gepList.emplace_back(gep);
-    m_chids[chid] = gepList;
-  }
-  
-  // Create the GEMChambers and store them on the Geometry 
-
-  for( const auto& ich : m_chids ) {
-    GEMDetId chid = ich.first;
-    std::list<GEMEtaPartition * > gepList = ich.second;
-
-    // compute the overall boundplane. At the moment we use just the last
-    // surface
-    BoundPlane* bp = nullptr;
-    for( const auto& gep : gepList ) {
-      const BoundPlane& bps = ( *gep ).surface();
-      bp = const_cast<BoundPlane *>( &bps );
+    if (gemid.roll() == 0){
+      if (gemid.layer() == 0){
+	GEMSuperChamber* gsc = buildSuperChamber( rgeo, id, gemid );
+	superChambers.insert(std::pair<GEMDetId, GEMSuperChamber*>(gemid, gsc));
+      }
+      else {
+	GEMChamber* gch = buildChamber( rgeo, id, gemid );
+	chambers.insert(std::pair<GEMDetId, GEMChamber*>(gemid, gch));
+      }
     }
-
-    ReferenceCountingPointer<BoundPlane> surf( bp );
-    // Create the superchamber
-    if( chid.layer() == 1 ) {
-      GEMDetId schid( chid.region(), chid.ring(), chid.station(), 0, chid.chamber(), 0 );
-      GEMSuperChamber* sch = new GEMSuperChamber( schid, surf );
-      LogDebug("GEMGeometryBuilder") << "GEM SuperChamber created with id = " << schid
-				     << " and added to the GEMGeometry" << std::endl;
-      superChambers.emplace_back( sch );
+    else {
+      GEMEtaPartition* gep = buildEtaPartition( rgeo, id, gemid );
+      partitions.insert(std::pair<GEMDetId, GEMEtaPartition*>(gemid, gep));
     }
     
-    // Create the chamber 
-    GEMChamber* ch = new GEMChamber( chid, surf ); 
-    LogDebug("GEMGeometryBuilder") << "GEM Chamber created with id = " << chid
-				   << " = " << chid.rawId() << " and added to the GEMGeometry" << std::endl;
-    LogDebug("GEMGeometryBuilder") << "GEM Chamber has following eta partitions associated: " << std::endl;
-
-    // Add the etaps to rhe chamber
-    for( const auto& gep : gepList ) {
-      ch->add(gep);
-      LogDebug("GEMGeometryBuilder") << "   --> GEM Eta Partition " << GEMDetId(( *gep ).id()) << std::endl;
-    }
-    // Add the chamber to the geometry
-    theGeometry->add( ch );
   }
-  
-  // The super chamber is composed of 2 chambers.
-  // It's detId is layer 0, chambers are layer 1 and 2
 
+  //auto& superChambers(theGeometry->superChambers());
   // construct the regions, stations and rings. 
-  for( int re = -1; re <= 1; re = re+2 ) {
-    GEMRegion* region = new GEMRegion( re );
-    for( int st = 1; st <= GEMDetId::maxStationId; ++st ) {
+  for (int re = -1; re <= 1; re = re+2){
+    GEMRegion* region = new GEMRegion(re);
```

Comment:
is this a memory leak? (I may have asked this before)

Reply:
this should be ok

Reply:
don't you have to delete the GEMRegion pointer after you add it to the GEMGeometry? Or use an appropriate smart pointer.